packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,218 @@
+    if (parent !== undefined) {
+      children = disconnectableChildrenForParent.get(parent)!;
+      children.delete(child);
+    } else {
+      break;
+    }
+  }
+};
+
+/**
+ * Sets the connected state on any directives contained within the committed
+ * value of this part (i.e. within a TemplateInstance or iterable of NodeParts)
+ * and runs their `disconnectedCallback`/`reconnectedCallback`s.
+ *
+ * `shouldRemoveFromParent` should be passed as `true` on a top-level part that
+ * is clearing itself, and not as a result of recursively disconnecting
+ * directives as part of a `clear` operation, as a performance optimization to
+ * avoid needless bookkeeping when a subtree is going away; when clearing a
+ * subtree, only the top-most part need to remove itself from the parent.
+ *
+ * Note, this method will be patched onto NodePart instances and called from the
+ * core code when parts are cleared or the connection state is changed by the
+ * user.
+ */
+function _setValueConnected(
+  this: NodePart,
+  isConnected: boolean,
+  shouldRemoveFromParent = false,
+  from = 0
+) {
+  const value = this._value;
+  if (value instanceof TemplateInstance) {
+    setConnected(value, isConnected, shouldRemoveFromParent);
+  } else if (Array.isArray(value)) {
+    for (let i = from; i < value.length; i++) {
+      setConnected(value[i], isConnected, shouldRemoveFromParent);
+    }
+  }
+  if (shouldRemoveFromParent) {
+    removeFromParentIfEmpty(this);
+  }
+}
+
+/**
+ * Sets the connected state on given directive if disconnectable and runs its
+ * `disconnectedCallback`/`reconnectedCallback`s.
+ *
+ * `shouldRemoveFromParent` should be passed as `true` only when a directive
+ * itself is disconnecting as a result of the part value changing, and not as a
+ * result of recursively disconnecting directives as part of a `clear`
+ * operation, as a performance optimization to avoid needless bookkeeping when a
+ * subtree is going away.
+ *
+ * Note, this method will be patched onto NodePart and AttributePart instances
+ * and called from the core code when directives are changed or removed.
+ */
+function _setDirectiveConnected(
+  this: Part,
+  directive: Directive | undefined,
+  isConnected: boolean,
+  shouldRemoveFromParent = false
+) {
+  if (directive instanceof DisconnectableDirective) {
+    directive._setConnected(isConnected);
+    if (shouldRemoveFromParent) {
+      disconnectableChildrenForParent.get(this)!.delete(directive);
+      removeFromParentIfEmpty(this);
+    }
+  }
+}
+
+/**
+ * TODO(kschaaf): Patches disconnection API onto the parent; we could also just
+ * install this on the prototype once, or bite the bullet and put it in core;
+ * the former would add a small runtime tax, and the latter would add a large
+ * code size tax.
+ */
+const installDisconnectAPI = (child: DisconnectableChild) => {
+  if (child instanceof AttributePart) {
+    child._setDirectiveConnected = _setDirectiveConnected;
```

Comment:
What happens if we get multiple copies of lit-html? I think installing of the prototype would be safer, because each prototype (for each copy of say, `NodePart`) would get the method installed once. Then if we avoid `instanceof` this should work across copies.